src/ExtranetBundle/LoginByTransferToken/TransferTokenService.php

```diff
@@ -0,0 +1,130 @@
+<?php
+
+namespace ChameleonSystem\ExtranetBundle\LoginByTransferToken;
+
+use ChameleonSystem\CoreBundle\Interfaces\TimeProviderInterface;
+use Psr\Log\LoggerInterface;
+
+class TransferTokenService implements TransferTokenServiceInterface
+{
+    private const DEFAULT_TOKEN = '!ThisTokenIsNotSoSecretChangeIt!';
+
+    /** @var TimeProviderInterface */
+    private $timeProvider;
+
+    /** @var LoggerInterface */
+    private $logger;
+
+    /** @var string */
+    private $secret;
+
+    /** @var string */
+    private $algorithm;
+
+    public function __construct(
+        TimeProviderInterface $timeProvider,
+        LoggerInterface $logger,
+        string $secret,
+        string $algorithm
+    ) {
+        $this->timeProvider = $timeProvider;
+        $this->logger = $logger;
+        $this->secret = $secret;
+        $this->algorithm = $algorithm;
+    }
+
+    /** {@inheritdoc} */
+    public function createTransferTokenForUser(
+        string $userId,
+        int $expiresAfterSeconds
+    ): string {
+        return $this->encodeToken([
+            'userId' => $userId,
+            'expires' => $this->timeProvider->getUnixTimestamp() + $expiresAfterSeconds
+        ]);
+    }
+
+    /** {@inheritdoc} */
+    public function getUserIdFromTransferToken(string $token): ?string
+    {
+       $data = $this->decodeToken($token);
+       if (null === $data
+           || false === array_key_exists('expires', $data)
+           || false === array_key_exists('userId', $data)
+       ) {
+           return null;
+       }
+
+        if ($this->timeProvider->getUnixTimestamp() > $data['expires']) {
+            return null;
+        }
+
+        return $data['userId'];
+    }
+
+    private function encodeToken(array $data): string
+    {
+        $data['_salt'] = md5(random_bytes(16));
```

Comment:
Why do secret things need to be md5 ?

Also in initializationVector().

Reply:
It doesn't have to be - the important part of this is converting the output of `random_bytes` into a regular string that does not have any weirdness in it. We don't have to do that, but doing so prevents future headaches when some invalid character is in the string. Basically what this does is generate a random hex-string.

We could use `base64encode` in these places - however, `md5` was chosen because it is a very well optimized implementation of this behavior.